apollo-router/src/query_planner/fetch.rs

```diff
@@ -243,6 +255,189 @@ impl Display for QueryHash {
+                            } else if !new_variables.iter().any(|var| var.name == v.name) {
+                                new_variables.push(v.clone());
+                            }
+                        }
+
+                        Ok(s)
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                let new_operation = operation.make_mut();
+                new_operation.selection_set = new_selection_set;
+                new_operation.variables = new_variables;
+
+                return Ok(Some(new_operation.serialize().no_indent().to_string()));
+            }
+        }
+    }
+
+    Ok(None)
+}
+
+fn rename_variables(selection_set: ast::Selection, from: Name, to: Name) -> ast::Selection {
+    match selection_set {
+        ast::Selection::Field(f) => {
+            let mut new = f.clone();
+
+            let as_mut = new.make_mut();
+
+            as_mut.arguments.iter_mut().for_each(|arg| {
+                if arg.value.as_variable() == Some(&from) {
+                    arg.make_mut().value = ast::Value::Variable(to.clone()).into();
+                }
+            });
+
+            as_mut.selection_set = as_mut
+                .selection_set
+                .clone()
+                .into_iter()
+                .map(|s| rename_variables(s, from.clone(), to.clone()))
+                .collect();
+
+            ast::Selection::Field(new)
+        }
+        ast::Selection::InlineFragment(f) => {
+            let mut new = f.clone();
+            new.make_mut().selection_set = f
+                .selection_set
+                .clone()
+                .into_iter()
+                .map(|s| rename_variables(s, from.clone(), to.clone()))
+                .collect();
+            ast::Selection::InlineFragment(new)
+        }
+        ast::Selection::FragmentSpread(f) => ast::Selection::FragmentSpread(f),
+    }
+}
+
+#[test]
+fn test_query_batching_for_contextual_args() {
+    let old_query = "query QueryLL__Subgraph1__1($representations:[_Any!]!$Subgraph1_U_field_a:String!){_entities(representations:$representations){...on U{id field(a:$Subgraph1_U_field_a)}}}";
+    let mut ctx_args = HashSet::new();
+    ctx_args.insert("Subgraph1_U_field_a".to_string());
+    let contextual_args = Some((ctx_args, 2));
+
+    let expected = "query QueryLL__Subgraph1__1($representations: [_Any!]!, $Subgraph1_U_field_a_0: String!, $Subgraph1_U_field_a_1: String!) { _0: _entities(representations: $representations) { ... on U { id field(a: $Subgraph1_U_field_a_0) } } _1: _entities(representations: $representations) { ... on U { id field(a: $Subgraph1_U_field_a_1) } } }";
+
+    assert_eq!(
+        expected,
+        query_batching_for_contextual_args(old_query, &contextual_args)
+            .unwrap()
+            .unwrap()
+    );
+}
+
+// TODO: There is probably a function somewhere else that already does this
+fn data_at_path<'v>(data: &'v Value, path: &Path) -> Option<&'v Value> {
```

Comment:
we have `select_values_and_paths`, or `get_path`, or the lower level `iterate_path`